docs/DesignDocs/NotesOnAttestationAPI.md

```diff
@@ -0,0 +1,160 @@
+Notes on OE SDK Attestation API and Plugin Library Design
+====
+
+This document contains notes on the concepts and assumptions behind the design
+of the OE SDK Attestation API as captured in document
+[CustomAttestation_V3.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation_V3.md),
+as well as the design of SGX plugins as described in document
+[Attestation: OE SDK Integration with Intel® SGX SDK quote-ex Library for Generation of Evidence in New Formats](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/SGX_QuoteEx_Integration.md).
+
+Hopefully these notes can help the readers' understanding of the API and its
+usage, and the SGX plugin design.
+
+# Semantics of Custom Claims and Optional Parameters
+
+To help clarify what goes into custom claims and what goes to optional
+parameters in the `oe_get_evidence()` API function, it would be good to
+articulate the semantics of these two fields.
+
+- Custom claims have well-defined semantics:
+  - Can be any name-value pair
+  - Only meaningful to ISV attester / verifier applications
+    - Typically, custom claims are part of an attester instance identity or
+    properties. They are created and maintained by the attester and not
+    received from a verifier
+    - Examples include a public key or a nonce generated by an attester
+    - These claims are passed to verifiers with integrity protection (as part
+    of the evidence), so verifiers can use the claims to learn the identity
+    and properties for the attester and for secure communication with the
+    latter
+  - They are always TEE agnostic
+  - Signed and included in evidence as clear text by an attester plugin,
+  and verified and reproduced by a verifier plugin, to ensure integrity
```

Comment:
Maybe "verified and restored"?

Reply:
My understanding of this passage is that the verifier will emulate whatever verification done in hardware (in SGX this could be recomputing the MRENCLAVE value or verifying the signature with a public key) to ensure the computed value matches the value reported by the attester.

I don't know how to say that in fewer words. Maybe that level of detail isn't even relevant for this document though. A verifier can verify evidence however it sees fit. Maybe that is computing a hash, maybe that is verifying a signature. The API probably doesn't need to be concerned with the how.

Reply:
@dthaler @yentsanglee yes "verified and reproduced" means "verified and restored", restoring to the original form of a list of name-value pairs for processing by verifier application. I'll edit this bullet.

@jhand2 SGX quote verification is quite complex, involving verification of quote and the cert chain that signs it. Though it does not involve re-computing MRENCLAVE. As you mentioned, how the evidence is verified is not the concern of this passage (which focuses on the parameters of the API.